File: app/src/test/java/com/example/recipeclipper/fake/FakeDecisionModel.kt
package com.example.recipeclipper.fake

import com.example.recipeclipper.data.DecisionModel
import com.example.recipeclipper.data.DecisionRepository
import com.example.recipeclipper.data.model.DecisionPrompt
import com.example.recipeclipper.data.model.DecisionQuestion
import com.example.recipeclipper.data.model.DecisionReply
import com.example.recipeclipper.data.model.Decisions
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow

/**
 * The decision model, faked (#104): [reply] answers each prompt (null: "can't right now"), by
 * default the same [answer] with confidence "high". [asked] records every prompt.
 */
class FakeDecisionModel(
    var languages: Set<String> = setOf("en", "de", "fr"),
    var answer: String = "unsure",
    var reply: (DecisionPrompt) -> DecisionReply? = { DecisionReply(answer, "high") }
) : DecisionModel {
    val asked = mutableListOf<DecisionPrompt>()

    override suspend fun supports(language: String) = language in languages

    override suspend fun ask(prompt: DecisionPrompt): DecisionReply? {
        asked += prompt
        return reply(prompt)
    }
}

/**
 * [DecisionRepository] in memory: [answers] are what the model would decide; [decide] moves
 * each asked question's answer into what [observe] emits. [asked] records the questions.
 */
class FakeDecisionRepository(val answers: Map<DecisionQuestion, String> = emptyMap()) : DecisionRepository {
    val asked = mutableListOf<DecisionQuestion>()
    val cached = MutableStateFlow<Map<DecisionQuestion, String>>(emptyMap())

    override fun observe(): Flow<Decisions> = decisions

    private val decisions = MutableStateFlow(Decisions.NONE)

    override suspend fun decide(questions: Collection<DecisionQuestion>) {
        asked += questions
        cached.value = cached.value + questions.mapNotNull { q -> answers[q]?.let { q to it } }
        decisions.value = Decisions(cached.value)
    }
}

File: ios/RecipeClipperTests/Fakes/FakeDecisionModel.swift
import Combine
import Foundation
@testable import RecipeClipper

/// The decision model, faked (#104): `reply` answers each prompt (nil: "can't right now"), by
/// default `answer` with confidence "high". `asked` records every prompt.
final class FakeDecisionModel: DecisionModel {
    var languages: Set<String>
    var answer: String
    var reply: ((DecisionPrompt) -> DecisionReply?)?
    private(set) var asked: [DecisionPrompt] = []

    init(languages: Set<String> = ["en", "de", "fr"], answer: String = "unsure") {
        self.languages = languages
        self.answer = answer
    }

    func supports(language: String) async -> Bool { languages.contains(language) }

    func ask(_ prompt: DecisionPrompt) async -> DecisionReply? {
        asked.append(prompt)
        if let reply { return reply(prompt) }
        return DecisionReply(answer: answer, confidence: "high")
    }
}

/// `DecisionRepository` in memory: `answers` are what the model would decide; `decide` moves
/// each asked question's answer into what `observe` emits. `asked` records the questions.
final class FakeDecisionRepository: DecisionRepository {
    let answers: [DecisionQuestion: String]
    private(set) var asked: [DecisionQuestion] = []
    private let subject = CurrentValueSubject<Decisions, Never>(.none)

    init(_ answers: [DecisionQuestion: String] = [:]) { self.answers = answers }

    func observe() -> AnyPublisher<Decisions, Never> { subject.eraseToAnyPublisher() }

    func current() async -> Decisions { subject.value }

    func decide(_ questions: [DecisionQuestion]) async {
        asked += questions
        var cached = subject.value.answers
        for q in questions { if let a = answers[q] { cached[q] = a } }
        subject.send(Decisions(answers: cached))
    }
}
